Added TeaCache Support and node

ctf05 opened the pull request:
It works. I did very little testing. You get a speed-up with a loss of quality but you can get a ~1.4x speedup with little loss.

DiffuzionDreamer commented:
I'm getting an error when running it in conjunction with Enhance-A-Video:
![Screenshot from 2024-12-31 18-10-22](https://github.com/user-attachments/assets/667bd817-ba4c-4bbb-aa83-e7f1b6358259)

But... I'm also able to run them together after running with just TeaCache and then again with both EAV and TC. Wouldn't know where to begin to fix that.

morganavr commented:
Sometimes after I reload workflow from .json I got this error.
I use TeaCache with Enhance-A-Video.
![image](https://github.com/user-attachments/assets/15a18e70-f1d1-4fa1-8627-12e97bef5b68)

DarioFT commented:
@DiffuzionDreamer I've created a small [PR](https://github.com/ctf05/ComfyUI-HunyuanVideoWrapper/pull/1) that addresses this on @ctf05 's fork.

ctf05 commented:
> @DiffuzionDreamer I've created a small [PR](https://github.com/ctf05/ComfyUI-HunyuanVideoWrapper/pull/1) that addresses this on @ctf05 's fork.

I just saw this. Thanks for the fix. It fixed some problems I had. I think TeaCache is bug free now. (known bugs)

DiffuzionDreamer commented:
> > @DiffuzionDreamer I've created a small [PR](https://github.com/ctf05/ComfyUI-HunyuanVideoWrapper/pull/1) that addresses this on @ctf05 's fork.
> 
> I just saw this. Thanks for the fix. It fixed some problems I had. I think TeaCache is bug free now. (known bugs)

Tested and working on my system, and seems to resolve the issue reported by @morganavr as well which had occured for me when aiming for larger/longer outputs.

morganavr commented:
TeaCache does not work with BlockSwap node - it throws an exception 
```
RuntimeError: Expected all tensors to be on the same device, but found at least two devices, cpu and cuda:0! (when checking argument for argument mat1 in method wrapper_CUDA_addmm)
```

![image](https://github.com/user-attachments/assets/bdd4731f-2f4e-4338-b7e4-f9387432898b)

kijai commented:
Thank you for this PR. I finally got my PC in working order again, so I took a look and implemented it slightly differently, I don't really want separate forward method that I'd have to keep up to date, so I added it in the existing model instead. This also allows it to work with all the other features like block swap and cpu offloading.

0000111100001111 commented:
I'm using full blockswap in a low memory environment. With the new dpm scheduler I don't see any reduction in generation times. With the original scheduler there's definitely a decrease. However, there's also a fairly large (relatively speaking) increase in vram usage... somewhere between .75-1.5gb I'd say. Is this to be expected?

kijai commented:
It looks like dpm is not compatible with TeaCache at all.

pondloso commented:
This is really game changer for hunyaun vid.
From my test it really much faster  at 0.15 is nearly take half time from with out it.
from 20 min to just 10-11 min but is clearly impact out put for me it look like 15-20 % down from without it
at 0.1 it more slower around 14 min but out put impact it better look 10% down from with out it.


but my best use is for preview i pump it to 0.4-0.5 it only take like 4 min to finish and look much more bad but surprisingly it not too different from 0.1 output so is good for preview before you decide with seed it the best.

It much much better than use low step to preview.

ctf05 commented:
> Thank you for this PR. I finally got my PC in working order again, so I took a look and implemented it slightly differently, I don't really want separate forward method that I'd have to keep up to date, so I added it in the existing model instead. This also allows it to work with all the other features like block swap and cpu offloading.

No problem, glad I could help.
